deps/rabbit/src/rabbit_db_queue.erl

```diff
@@ -409,17 +416,26 @@ delete_in_khepri(QueueName) ->
 delete_in_khepri(QueueName, OnlyDurable) ->
     rabbit_khepri:transaction(
       fun () ->
-              Path = khepri_queue_path(QueueName),
-              case khepri_tx_adv:delete(Path) of
-                  {ok, #{data := _}} ->
-                      %% we want to execute some things, as decided by rabbit_exchange,
-                      %% after the transaction.
-                      rabbit_db_binding:delete_for_destination_in_khepri(QueueName, OnlyDurable);
-                  {ok, _} ->
-                      ok
-              end
+              delete_in_khepri_tx(QueueName, OnlyDurable)
       end, rw).
 
+delete_in_khepri_tx(QueueName, OnlyDurable) ->
+    Path = khepri_queue_path(QueueName),
+    case khepri_tx_adv:delete(Path) of
+        %% Khepri 0.16 and below returned `khepri:node_props()' for
+        %% adv queries and commands targeting one node:
+        {ok, #{data := _}} ->
+            %% we want to execute some things, as decided by rabbit_exchange,
+            %% after the transaction.
+            rabbit_db_binding:delete_for_destination_in_khepri(QueueName, OnlyDurable);
+        %% Khepri 0.17+ returns `khepri_adv:node_props_map()`
+        %% instead.
+        {ok, #{Path := #{data := _}}} ->
+            rabbit_db_binding:delete_for_destination_in_khepri(QueueName, OnlyDurable);
```

Comment:
Oh, that’s a good point! This probably needs a fix in Khepri. I see two solutions:
1. The typespec is updated to document the old return value (which it can return)
2. The function converts the old return value to the new one (I didn’t check if it was possible though)

What do you think?

Reply:
Do we need a change in Khepri? https://github.com/rabbitmq/khepri/pull/303 only touched the "client" code (i.e. not `khepri_machine`) so older versions should continue to see `khepri:node_props()` while newer versions consistently see `khepri_adv:node_props_map()`, even if the leader were an older versioned node.

The typespec change is unfortunate and the dialyzer had a few complaints that I worked around with `-dialyzer({nowarn_function, [..]}).` attributes. I think this is ok for compatibility code like this. We should be able to delete these branches and remove the dialyzer ignores for 4.2.x once we don't need compatibility with Khepri 0.16. (Assuming Khepri 0.17 would ship with 4.1.x.)

Reply:
> newer versions consistently see khepri_adv:node_props_map(), even if the leader were an older versioned node.

Right, we have `reply_from => LocalMember` for commands, so the return value is created by the local version of Khepri. Then, why do we need to have the `case` for both return values?

With the discussions around long-term support for RabbitMQ, compatibility code can stay for a long time as we will have to support upgrades that skip versions. That said, it might not be the case with RabbitMQ 4.0.x. This is more of a general comment.

Reply:
Oh actually that's a really good point, I suppose the old return value isn't possible and we shouldn't need this compatibility code